mypy/messages.py

```diff
@@ -119,12 +127,18 @@ def report(self, msg: str, context: Optional[Context], severity: str,
                                context.get_column() if context else -1,
                                msg, severity=severity, file=file, offset=offset,
                                origin_line=origin.get_line() if origin else None,
-                               end_line=end_line)
-
-    def fail(self, msg: str, context: Optional[Context], file: Optional[str] = None,
+                               end_line=end_line,
+                               code=code)
+
+    def fail(self,
+             msg: str,
+             context: Optional[Context],
+             *,
+             code: Optional[ErrorCode] = None,
```

Comment:
We should add the `code` argument to the versions of this method in `SemanticAnalyzerCoreInterface` and maybe in `TypeAnalyzerPluginInterface`.